adapter/src/main/java/com/twilio/kudu/sql/rules/KuduSortRule.java

```diff
@@ -77,7 +79,16 @@ public boolean canApply(final RelTraitSet sortTraits, final KuduQuery query, fin
     final RelCollation collation = sortTraits.getTrait(RelCollationTraitDef.INSTANCE);
 
     if (collation.getFieldCollations().isEmpty()) {
-      return false;
+      if (offset != null) {
+        // if we have an OFFSET return rows sorted by the primary key even if there
+        // isn't an
+        // ORDER BY clause so rows are returned in a deterministic order
+        pkSortColumns = IntStream.range(0, openedTable.getSchema().getPrimaryKeyColumnCount()).boxed()
+            .collect(Collectors.toList());
+        return true;
+      } else {
+        return false;
+      }
```

Comment:
on the fence FWIW. Using Sorted Enumerable does make the query execute slowly.

Reply:
This is the existing behavior, this just makes it explicit by ensuring that we see that the rows are returned in sorted order if the query has an OFFSET. 
https://github.com/twilio/calcite-kudu/blob/71991a984e0c7357a2d0909741da675b264a65b7/adapter/src/main/java/com/twilio/kudu/sql/KuduEnumerable.java#L147

Reply:
O well ! k